pkg/controller/nodeipam/ipam/node_topology_syncer.go

```diff
@@ -0,0 +1,228 @@
+	if node == nil || err != nil {
+		klog.InfoS("Node not found or error, reconcile.", "node key", key, "error", err)
+		err := syncer.reconcile()
+		if err != nil {
+			klog.ErrorS(err, "Failed to reconcile nodeTopology CR")
+			return err
+		}
+	} else {
+		err := syncer.updateNodeTopology(node)
+		if err != nil {
+			klog.ErrorS(err, "Failed to add or update nodeTopology CR")
+			return err
+		}
+
+	}
+	return nil
+}
+
+func (syncer *NodeTopologySyncer) reconcile() error {
+	if syncer.nodeLister == nil {
+		return nil
+	}
+	allNodes, err := syncer.nodeLister.List(labels.NewSelector())
+	if err != nil {
+		klog.ErrorS(err, "Failed to list all nodes from nodeInformer lister")
+		return err
+	}
+
+	defaultSubnet, subnetPrefix, err := getSubnetWithPrefixFromURL(syncer.cloud.SubnetworkURL())
+	if err != nil {
+		klog.ErrorS(err, "Error parsing the default subnetworkURL")
+		return err
+	}
+	updatedSubnetsMap := make(map[string]nodetopologyv1.SubnetConfig, 0)
+	for _, node := range allNodes {
+		hasSubnetLabel, nodeSubnet := getNodeSubnetLabel(node)
+		if hasSubnetLabel {
+			updatedSubnetsMap[nodeSubnet] = nodetopologyv1.SubnetConfig{
+				Name:       nodeSubnet,
+				SubnetPath: subnetPrefix + nodeSubnet,
+			}
+			klog.InfoS("Making node topology subnets list for all nodes for subnet", "subnet", nodeSubnet)
+		}
+	}
+	updatedSubnetsMap[defaultSubnet] = nodetopologyv1.SubnetConfig{
+		Name:       defaultSubnet,
+		SubnetPath: subnetPrefix + defaultSubnet,
+	}
+
+	nodeTopologyCR, err := syncer.nodeTopologyClient.NetworkingV1().NodeTopologies().Get(context.TODO(), nodeTopologyCRName, metav1.GetOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Failed to get NodeTopology", "nodeTopologyCR", nodeTopologyCRName)
+		return err
+	}
+	updatedNodeTopologyCR := nodeTopologyCR.DeepCopy()
+	updatedSubnets := make([]nodetopologyv1.SubnetConfig, 0)
+	for _, s := range updatedSubnetsMap {
+		updatedSubnets = append(updatedSubnets, s)
+	}
+	updatedNodeTopologyCR.Status.Subnets = updatedSubnets
+
+	if updatedNodeTopologyCR.Status.Zones == nil {
+		updatedNodeTopologyCR.Status.Zones = []string{}
+	}
+	_, updateErr := syncer.nodeTopologyClient.NetworkingV1().NodeTopologies().UpdateStatus(context.TODO(), updatedNodeTopologyCR, metav1.UpdateOptions{})
+	if updateErr != nil {
+		klog.ErrorS(updateErr, "Error updating nodeTopology CR", "nodetopologyCR", nodeTopologyCRName)
+		return updateErr
+	}
+	klog.InfoS("Successfully reconciled nodeTopolody CR")
+	return nil
+}
+
+func (syncer *NodeTopologySyncer) updateNodeTopology(node *v1.Node) error {
+	hasSubnetLabel, nodeSubnet := getNodeSubnetLabel(node)
+
+	defaultSubnet, subnetPrefix, err := getSubnetWithPrefixFromURL(syncer.cloud.SubnetworkURL())
+	if err != nil {
+		klog.Errorf("Error parsing the default subnetworkURL, err: %v", err)
+		return err
+	}
+
+	nodeTopologyCR, err := syncer.nodeTopologyClient.NetworkingV1().NodeTopologies().Get(context.TODO(), nodeTopologyCRName, metav1.GetOptions{})
```

Comment:
what will happen if the CR is not present?
I'm guessing you will get a NotFound error here, I'm wondering if we should somehow handle this? I see that it should be mainained by the addon manager so maybe it's ok here, it should be re-added.

Reply:
There will be error and the entry will re-queued.